File: AgencyCampaign/AgencyCampaign.Web/src/services/integrationService.ts
import { httpClient } from 'archon-ui'
import type { Integration } from '../types/integration'

const BASE_URL = '/Integrations'

export interface CreateIntegrationRequest {
  identifier: string
  name: string
  description?: string
  categoryId: number
}

export interface UpdateIntegrationRequest extends CreateIntegrationRequest {
  id: number
  isActive: boolean
}

function extractItems<T>(data: T[] | { items?: T[] } | undefined): T[] {
  return Array.isArray(data) ? data : data?.items ?? []
}

export const integrationService = {
  async getAll(): Promise<Integration[]> {
    const response = await httpClient.get<Integration[] | { items?: Integration[] }>(`${BASE_URL}/Get`)
    return extractItems(response.data)
  },

  async getActive(): Promise<Integration[]> {
    const response = await httpClient.get<Integration[]>(`${BASE_URL}/active`)
    return response.data ?? []
  },

  async getById(id: number): Promise<Integration | null> {
    const response = await httpClient.get<Integration>(`${BASE_URL}/${id}`)
    return response.data ?? null
  },

  create(data: CreateIntegrationRequest) {
    return httpClient.post<Integration>(`${BASE_URL}/Create`, data)
  },

  update(id: number, data: UpdateIntegrationRequest) {
    return httpClient.put<Integration>(`${BASE_URL}/Update/${id}`, data)
  },
}

File: AgencyCampaign/AgencyCampaign.Infrastructure/Services/IntegrationService.cs
using AgencyCampaign.Application.Localization;
using AgencyCampaign.Application.Requests.Integrations;
using AgencyCampaign.Application.Services;
using AgencyCampaign.Domain.Entities;
using Archon.Core.Pagination;
using Archon.Infrastructure.Persistence.EF;
using Archon.Infrastructure.Services;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Localization;

namespace AgencyCampaign.Infrastructure.Services
{
    public sealed class IntegrationService : CrudService<Integration>, IIntegrationService
    {
        private readonly IStringLocalizer<AgencyCampaignResource> localizer;

        public IntegrationService(DbContext dbContext, IStringLocalizer<AgencyCampaignResource> localizer) : base(dbContext)
        {
            this.localizer = localizer;
        }

        public async Task<PagedResult<Integration>> GetIntegrations(PagedRequest request, CancellationToken cancellationToken = default)
        {
            return await DbContext.Set<Integration>()
                .AsNoTracking()
                .OrderBy(item => item.Name)
                .ToPagedResultAsync(request, cancellationToken);
        }

        public async Task<List<Integration>> GetActiveIntegrations(CancellationToken cancellationToken = default)
        {
            return await DbContext.Set<Integration>()
                .AsNoTracking()
                .Where(item => item.IsActive)
                .OrderBy(item => item.Name)
                .ToListAsync(cancellationToken);
        }

        public async Task<Integration?> GetIntegrationById(long id, CancellationToken cancellationToken = default)
        {
            return await DbContext.Set<Integration>()
                .AsNoTracking()
                .FirstOrDefaultAsync(item => item.Id == id, cancellationToken);
        }

        public async Task<Integration> CreateIntegration(CreateIntegrationRequest request, CancellationToken cancellationToken = default)
        {
            Integration integration = new(request.Identifier, request.Name, request.CategoryId, request.Description);
            bool success = await Insert(cancellationToken, integration);
            if (!success)
            {
                throw new InvalidOperationException(GetErrorMessages());
            }

            return integration;
        }

        public async Task<Integration> UpdateIntegration(long id, UpdateIntegrationRequest request, CancellationToken cancellationToken = default)
        {
            if (id != request.Id)
            {
                throw new InvalidOperationException(localizer["request.route.idMismatch"]);
            }

            Integration? integration = await DbContext.Set<Integration>()
                .AsTracking()
                .FirstOrDefaultAsync(item => item.Id == id, cancellationToken);

            if (integration is null)
            {
                throw new InvalidOperationException(localizer["record.notFound"]);
            }

            integration.Update(request.Identifier, request.Name, request.CategoryId, request.Description, request.IsActive);

            Integration? result = await Update(integration, cancellationToken);
            if (result is null)
            {
                throw new InvalidOperationException(GetErrorMessages());
            }

            return result;
        }
    }
}
